intermediate_source/named_tensor_tutorial.py

```diff
@@ -0,0 +1,519 @@
+# (``tensor.align_to(*names)`` works as well):
+
+imgs = imgs.refine_names('N', 'C', 'H', 'W')
+per_batch_scale = per_batch_scale.refine_names('N')
+
+named_result = imgs * per_batch_scale.align_as(imgs)
+assert torch.allclose(named_result.rename(None), correct_result)
+
+######################################################################
+# New behavior: Flattening and unflattening dimensions by names
+# -------------------------------------------------------------
+#
+# One common operation is flattening and unflattening dimensions. Right now,
+# users perform this using either ``view``, ``reshape``, or ``flatten``; use
+# cases include flattening batch dimensions to send tensors into operators that
+# must take inputs with a certain number of dimensions (i.e., conv2d takes 4D input).
+#
+# To make these operation more semantically meaningful than view or reshape, we
+# introduce a new ``tensor.unflatten(dim, namedshape)`` method and update
+# ``flatten`` to work with names: ``tensor.flatten(dims, new_dim)``.
+#
+# ``flatten`` can only flatten adjacent dimensions but also works on
+# non-contiguous dims. One must pass into ``unflatten`` a **named shape**, which
+# is a list of ``(dim, size)`` tuples, to specify how to unflatten the dim. It
+# is possible to save the sizes during a ``flatten`` for ``unflatten`` but we
+# do not yet do that.
+
+imgs = imgs.flatten(['C', 'H', 'W'], 'features')
+print(imgs.names)
+
+imgs = imgs.unflatten('features', (('C', 2), ('H', 2), ('W', 2)))
+print(imgs.names)
+
+######################################################################
+# Autograd support
+# ----------------
+#
+# Autograd currently supports named tensors in a limited manner: autograd
+# ignores names on all tensors. Gradient computation is still correct but we
+# lose the safety that names give us. It is on the roadmap to introduce handling
+# of names to autograd.
+
+x = torch.randn(3, names=('D',))
+weight = torch.randn(3, names=('D',), requires_grad=True)
+loss = (x - weight).abs()
+grad_loss = torch.randn(3)
+loss.backward(grad_loss)
+
+print(weight.grad)  # Unnamed for now. Will be named in the future
+
+weight.grad.zero_()
+grad_loss = grad_loss.refine_names('C')
+loss = (x - weight).abs()
+# Ideally we'd check that the names of loss and grad_loss match but we don't yet.
+loss.backward(grad_loss)
+
+print(weight.grad)  # still unnamed
+
+######################################################################
+# Other supported (and unsupported) features
+# ------------------------------------------
+#
+# See here (link to be included) for a detailed breakdown of what is
+# supported with the 1.3 release, what is on the roadmap to be supported soon,
+# and what will be supported in the future but not soon.
+#
+# In particular, we want to call out three important features that are not
+# currently supported:
+#
+# - Retaining names when serializing or loading a serialized ``Tensor`` via

```

Comment:
"Retaining names when ..." is a little misleading because it implies that `torch.save` will work but just not retain names. `torch.save` errors out gracefully when passed a named tensor